README.md
# OnlineSubsystemEpic
An unofficial integration of the Epic Online Systems SDK with Unreal Engine 4's OnlineSubsystem

### Currently Supported Versions
The plugin is coded against a specific version of the EOS SDK. While Epic promises to deliver a stable ABI, it's best to use the Plugin with the SDK version it was coded against

|   Plugin Version   | SDK Version |
| ------------------ |------------ |
|       0.0.3        |    1.7.0    |
|       0.0.2        |    1.6.0    |
|       0.0.1        |    1.6.0    |


## Implemented Interfaces
Use this section to see which functions of the EOS SDK are already implemented. For some interfaces it might be, that they are not completely mapped to the SDK functions, either because the SDK doesn't support the functionality, or the OSS doesn't. There are three symbols, you can use to check if the interface meets your expectations:
* :heavy_check_mark: - The interface is completely implemented and works as the OSS documentations describes.
* :warning: - Parts of the interface are implemented. See the notes field for additional information
* :x: - The interface isn't implemented

Currently there exists 36 interfaces in the online subsystem, of which 26 are publicly available. These interfaces and their implementation status are

|  Subsystem Interface  | Implementation Status |       Additional Notes       |
| --------------------- | --------------------- | ---------------------------- |
| OnlineSession         |       :warning:       | See [Known Limitations](KnownLimitations.md) |
| OnlineFriends         |         :x:           |
| OnlineParty           |         :x:           |
| OnlineGroups          |         :x:           |
| OnlineSharedCloud     |         :x:           |
| OnlineUserCloud       |         :x:           |
| OnlineEntitlements    |         :x:           |
| OnlineLeaderboards    |         :x:           |
| OnlineVoice           |         :x:           |
| OnlineExternalUI      |         :x:           |
| OnlineTime            |         :x:           |
| OnlineIdentity        |       :warning:       | See [Known Limitations](KnownLimitations.md) |
| OnlineTitleFile       |         :x:           |
| OnlineStore           |         :x:           |
| OnlineStoreV2         |         :x:           |
| OnlinePurchase        |         :x:           |
| OnlineEvents          |         :x:           |
| OnlineAchievements    |         :x:           |
| OnlineSharing         |         :x:           |
| OnlineUser            |       :warning:       | See [Known Limitations](KnownLimitations.md) |
| OnlineMessage         |         :x:           |
| OnlinePresence        |       :warning:       | See [Known Limitations](KnownLimitations.md) |
| OnlineChat            |         :x:           |
| OnlineStats           |         :x:           |
| OnlineTurnBased       |         :x:           |
| OnlineTournament      |         :x:           |


## Installation
1. Clone the repository into your Projects _Plugins_ directory.
2. Download the EOS C++ SDK from the Developer Portal
3. Drop the _Bin_, _Include_ and _Lib_ from inside the SDK folder into _Source\ThirdParty\OnlineSubsystemEpicLibrary_

To make the SDK initialize as intended a few properties have to be set. As not to hardcode them, they are set inside two .ini files, which can easily be changed without a recompile. Keep in mind, that some of these properties contain sensitive information and **mustn't** be made public. Therefore it's best to keep it out of source control. Later versions might include different retrieval mechanisms.

DefaultGame.ini
```ini
[/Script/EngineSettings.GeneralProjectSettings]
ProjectID = <YourProjectID>
ProjectName = <YourProjectName>
ProjectVersion = <YourProjectVersion>
```
DefaultEngine.ini
```ini
[OnlineSubsystem]
DefaultPlatformService=Epic

[OnlineSubsystemEpic]
; The product id for the running application, found on the dev portal
ProductId = <ProductId>
; The sandbox id for the running application, found on the dev portal
SandboxId = <SandboxId>
; The deployment id for the running application, found on the dev portal
DeploymentId = <DeploymentId>
; Client id of the service permissions entry
ClientCredentialsId = <ClientCredentialsId>
; Client secret for accessing the set of permissions
ClientCredentialsSecret = <ClientCredentialsSecret>

```

Additionally the SDK supports a a few optional settings for further customization:
DefaultEngine.ini
```ini
; Overrides the internal country code, max 4 characters in length.
CountryCode = <ISO3166CountryCode>
; Overrides the locale used by the SDK, max 9 characters in length
LocaleCode = <ISO639LanguageCode>
; Encryption key, 64 hex characters long. Used only by player data storage
EncryptionKey = <Custom64CharacterLongEncryptionKey>
; Sets the directory used for SDK side caching. The path is created if missing
CacheDirectory = <AbsoluteCachePath>
; A budget, measured in milliseconds, for ticking tasks to do their work.
; When the budget is met or exceeded (or if no work available), the tasks will return.
; This allows the SDK to amortize the cost of work across multiple frames in the event that a lot of work is queued for processing.
; Zero is interpreted as "perform all available work"
; The codomain of this variable the same as an unsigned 32-bit integer,
; although it's represented as a double in this config, since UE4 doesn't allow
; retrieval of unsigned 32-bit integers from code.
TickBudget = <DurationInMs>
; Indicates the SDK should skip initialization of the overlay, used by the in-app purchase flow and social overlay.
; Implied by LoadingInEditor
DisableOverlay = <true>/<false>
; Indicates the SDK should skip initialization of the social overlay.
; Implied by DisableOverlay
DisableSocialOverlay = <true>/<false>
; Change if the Developer Auth Tool doesn't live on the local machine
; or the port 9999 is not available. Default: 127.0.0.1:9999
DevToolAddress=<IPv4 or IPv6 Address>
```

## Usage
This plugin is used like any other OnlineSubsystem Plugin already existing. This means, that most of the time you won't need to directly interface with the system directly, but can let the engine classes handle the calls.
If you need to directly access the OnlineSubsystem you should get it via the static helper methods in `Online.h`. These helper methods make sure the correct subsystem instance is retrieved (multiple can exist in the editor, and things like logins are tied to a specific instance). Outside of C++ there exists multiple asynchronous blueprint nodes in the _OnlineSubsystemUtils_ plugin. In most cases there is no need to access the online subsystem via `IOnlineSubsystem::Get()`.

### Identity Interface
The identity interface is the central hub for access management. It provides the ability to login and logout a user, check their login status and get their player ids.
This plugin supports two login flows: One, an *Open Id Connect* (Connect) compliant, as well as Epics own account system (EAS) login flow. With this there are some things a user has to consider when logging in a user.
User credentials are passed to the system via the `FOnlineAccountCredentials` class. This class has a `Type` field, which is used to distingush between EAS and Connect.
To switch between the two, the `Type` field has the format `{FlowType}:{LoginType}` where the `FlowType` is either "EAS" or "CONNECT" and `LoginType` one of the following:

|       EAS      |    Connect    |
| -------------- | ------------- |
| Password       | Steam         |
| ExchangeCode   | PSN           |
| PersistentAuth | XBL           |
| DeviceCode     | Discord       |
| Developer      | GOG           |
| RefreshToken   | Nintendo_Id   |
| AccountPortal  | Nintentod_NSA |
|                | Uplay         |
|                | OpenId        |
|                | DeviceId      |
|                | Apple         |

While connect also supports EAS as `LoginType`, this is implicitly assumed by this plugin when using the EAS login flow.

`FOnlineAccountCredentials` class has two additional fields, `Id` and `Token`. When using "CONNECT" login flow the `Token` field stores the access token, while the `Id` field holds additional data, that is needed for the Nintendo and Apple login types.
When using "EAS" as login flow, consult the "OnlineIdentityInterface.h" file to see which field maps to which.

#### Continuance Tokens
**Note::** This is currently not supported as the EOS SDK gives no possibility to convert a continuance token to and from strings.

When using the connect interface, there might not be a user to login with. The interface remedies that, that it return a continuance token with which the caller can restart the login process. This library supports this in multiple ways.
In *C++* the OnLoginCompleteDelegate is called regardless if the task completed successful or not. If the original call completed without errors, the delegate will have set the `bWasSuccessful` parameter set to `true` and will contain the local user index, and the users unique net id. If the user doesn't exist but the login process can be restarted by using a continuance token the `bWasSuccessful` parameter is set to `false` and the unique net id will contain the *continuance token*. The process then can be restarted by calling the `IOnlineIdentityInterface::Login(int32, const FOnlineAccountCredentials&)` function, where the `FOnlineAccountCredentials` parameter is initialized with the following parameters:
* Id: *empty*
* Token: `<continuance token>`
* Type: CONNECT::Continuance

If this call succeeds the OnLoginCompleteDelegate is called with the `bWasSuccessful` parameter set to `true`, the local user index, and the users unique net id. Note, that since a continuance token was used to create an account, the unique net id will only contain a product user id and never an epic account id.
Should the call fail, the delegate will be called with the local user index, the `bWasSuccessful` parameter set to `false`, an invalid net id and an error message.

In *Blueprints* the caller doesn't need to do anything. The BP-Node will take the login details and a boolean asking whether to create a new user. The node then will internally call the appropriate C++ functions.


CHANGELOG.md
# Changelog

## [Unreleased](https://github.com/Ruhrpottpatriot/OnlineSubsystemEpic/tree/HEAD)

[Full Changelog](https://github.com/Ruhrpottpatriot/OnlineSubsystemEpic/compare/v0.3_Presence...HEAD)

**Breaking changes:**

- Remove ability to create an FUniqueNetIdEpic from string [\#34](https://github.com/Ruhrpottpatriot/OnlineSubsystemEpic/issues/34)
- Add support for connect login and support for external account creation [\#29](https://github.com/Ruhrpottpatriot/OnlineSubsystemEpic/pull/29) ([Ruhrpottpatriot](https://github.com/Ruhrpottpatriot))
- Make address of DevAuthTool customizable [\#24](https://github.com/Ruhrpottpatriot/OnlineSubsystemEpic/pull/24) ([Ruhrpottpatriot](https://github.com/Ruhrpottpatriot))

**Implemented enhancements:**

- Change FUniqueNetIdEpic::GetHash [\#36](https://github.com/Ruhrpottpatriot/OnlineSubsystemEpic/issues/36)
- Allow an FUniqueNetIdEpic to be created from an Epic Account ID [\#33](https://github.com/Ruhrpottpatriot/OnlineSubsystemEpic/issues/33)
- DevAuthTool should have the address customizable [\#23](https://github.com/Ruhrpottpatriot/OnlineSubsystemEpic/issues/23)

**Fixed bugs:**

- Session Ids are still created as FUniqueNetIdEpic  [\#35](https://github.com/Ruhrpottpatriot/OnlineSubsystemEpic/issues/35)
- FOnlineSessionEpic::FindSessions doesn't correctly cast from FUniqueNetId to FUniqueNetIdEpic [\#26](https://github.com/Ruhrpottpatriot/OnlineSubsystemEpic/issues/26)
- OnlineSubsystemEpic tries to initialize missing devToolPort field [\#25](https://github.com/Ruhrpottpatriot/OnlineSubsystemEpic/issues/25)

## [v0.3_Presence](https://github.com/Ruhrpottpatriot/OnlineSubsystemEpic/tree/v0.3_Presence) (2020-07-02)

[Full Changelog](https://github.com/Ruhrpottpatriot/OnlineSubsystemEpic/compare/v0.2_Sessions...v0.3_Presence)

**Implemented enhancements:**

- Make FUniqueNetIdEpic a proper type [\#7](https://github.com/Ruhrpottpatriot/OnlineSubsystemEpic/issues/7)
- Add support for EOS\_Connect [\#20](https://github.com/Ruhrpottpatriot/OnlineSubsystemEpic/issues/20)
- Change FOnlineSessionEpic::CreateSessionIdFromString to use FUniqueNetIdString [\#19](https://github.com/Ruhrpottpatriot/OnlineSubsystemEpic/issues/19)
- GetAppId should return project version [\#18](https://github.com/Ruhrpottpatriot/OnlineSubsystemEpic/issues/18)
- Add LoginType to FUniqueNetIdEpic [\#9](https://github.com/Ruhrpottpatriot/OnlineSubsystemEpic/pull/9) ([Ruhrpottpatriot](https://github.com/Ruhrpottpatriot))

**Fixed bugs:**

- 4 errors on first install and compile. [\#12](https://github.com/Ruhrpottpatriot/OnlineSubsystemEpic/issues/12)
- Compile error with latest commit [\#6](https://github.com/Ruhrpottpatriot/OnlineSubsystemEpic/issues/6)

**Merged pull requests:**

- Fix compile error on linux [\#13](https://github.com/Ruhrpottpatriot/OnlineSubsystemEpic/pull/13) ([Qyg0n](https://github.com/Qyg0n))
- Adds async task blueprint nodes and fixes some bugs in the identity interface [\#5](https://github.com/Ruhrpottpatriot/OnlineSubsystemEpic/pull/5) ([CNDW](https://github.com/CNDW))

## [v0.2_Sessions](https://github.com/Ruhrpottpatriot/OnlineSubsystemEpic/tree/v0.2_Sessions) (2020-06-05)

[Full Changelog](https://github.com/Ruhrpottpatriot/OnlineSubsystemEpic/compare/v0.1_Identity...v0.2_Sessions)

## [v0.1_Identity](https://github.com/Ruhrpottpatriot/OnlineSubsystemEpic/tree/v0.1_Identity) (2020-05-31)

[Full Changelog](https://github.com/Ruhrpottpatriot/OnlineSubsystemEpic/compare/8b42b9d3f047675181bf70758d49541d40deb22d...v0.1_Identity)

**Implemented enhancements:**

- Allow dev tool port to be set in config files [\#22](https://github.com/Ruhrpottpatriot/OnlineSubsystemEpic/issues/22)
- adds support for autologin method [\#2](https://github.com/Ruhrpottpatriot/OnlineSubsystemEpic/pull/2) ([CNDW](https://github.com/CNDW))

**Fixed bugs:**

- Disables the platform overlay in the editor and PIE to fix editor rendering issues [\#1](https://github.com/Ruhrpottpatriot/OnlineSubsystemEpic/pull/1) ([CNDW](https://github.com/CNDW))

**Closed issues:**

- Repository should contain a file with known limitations [\#21](https://github.com/Ruhrpottpatriot/OnlineSubsystemEpic/issues/21)



\* *This Changelog was automatically generated by [github_changelog_generator](https://github.com/github-changelog-generator/github-changelog-generator)*


Source/OnlineSubsystemEpic/Private/OnlineSubsystemEpic.cpp
#include "OnlineSubsystemEpic.h"
#include "OnlineIdentityInterfaceEpic.h"
#include "OnlineSessionInterfaceEpic.h"
#include "OnlineUserInterfaceEpic.h"
#include "Utilities.h"
#include <string>
#include "OnlineFriendInterfaceEpic.h"

IOnlineSessionPtr FOnlineSubsystemEpic::GetSessionInterface() const
{
	return this->SessionInterface;
}

IOnlineFriendsPtr FOnlineSubsystemEpic::GetFriendsInterface() const
{
	return this->FriendsInterface;
}

IOnlinePartyPtr FOnlineSubsystemEpic::GetPartyInterface() const
{
	return nullptr;
}

IOnlineGroupsPtr FOnlineSubsystemEpic::GetGroupsInterface() const
{
	return nullptr;
}

IOnlineSharedCloudPtr FOnlineSubsystemEpic::GetSharedCloudInterface() const
{
	return nullptr;
}

IOnlineUserCloudPtr FOnlineSubsystemEpic::GetUserCloudInterface() const
{
	return nullptr;
}

IOnlineEntitlementsPtr FOnlineSubsystemEpic::GetEntitlementsInterface() const
{
	return nullptr;
}

IOnlineLeaderboardsPtr FOnlineSubsystemEpic::GetLeaderboardsInterface() const
{
	return nullptr;
}

IOnlineVoicePtr FOnlineSubsystemEpic::GetVoiceInterface() const
{
	return nullptr;
}

IOnlineExternalUIPtr FOnlineSubsystemEpic::GetExternalUIInterface() const
{
	return nullptr;
}

IOnlineTimePtr FOnlineSubsystemEpic::GetTimeInterface() const
{
	return nullptr;
}

IOnlineIdentityPtr FOnlineSubsystemEpic::GetIdentityInterface() const
{
	return this->IdentityInterface;
}

IOnlineTitleFilePtr FOnlineSubsystemEpic::GetTitleFileInterface() const
{
	return nullptr;
}

IOnlineStorePtr FOnlineSubsystemEpic::GetStoreInterface() const
{
	return nullptr;
}

IOnlineStoreV2Ptr FOnlineSubsystemEpic::GetStoreV2Interface() const
{
	return nullptr;
}

IOnlinePurchasePtr FOnlineSubsystemEpic::GetPurchaseInterface() const
{
	return nullptr;
}

IOnlineEventsPtr FOnlineSubsystemEpic::GetEventsInterface() const
{
	return nullptr;
}

IOnlineAchievementsPtr FOnlineSubsystemEpic::GetAchievementsInterface() const
{
	return nullptr;
}

IOnlineSharingPtr FOnlineSubsystemEpic::GetSharingInterface() const
{
	return nullptr;
}

IOnlineUserPtr FOnlineSubsystemEpic::GetUserInterface() const
{
	return this->UserInterface;
}

IOnlineMessagePtr FOnlineSubsystemEpic::GetMessageInterface() const
{
	return nullptr;
}

IOnlinePresencePtr FOnlineSubsystemEpic::GetPresenceInterface() const
{
	return this->PresenceInterface;
}

IOnlineChatPtr FOnlineSubsystemEpic::GetChatInterface() const
{
	return nullptr;
}

IOnlineStatsPtr FOnlineSubsystemEpic::GetStatsInterface() const
{
	return nullptr;
}

IOnlineTurnBasedPtr FOnlineSubsystemEpic::GetTurnBasedInterface() const
{
	return nullptr;
}

IOnlineTournamentPtr FOnlineSubsystemEpic::GetTournamentInterface() const
{
	return nullptr;
}

bool FOnlineSubsystemEpic::Init()
{
	if (this->IsInit)
	{
		UE_LOG_ONLINE(Warning, TEXT("Tried initializing already initialized subsystem"));
		return false;
	}
	bool hasInvalidParams = false;

	// Get the Developer tool port.
	// If this isn't set, we default to 9999
	if (!GConfig->GetString(
		TEXT("OnlineSubsystemEpic"),
		TEXT("DevToolAddress"),
		this->DevToolAddress,
		GEngineIni))
	{
		UE_LOG_ONLINE(Verbose, TEXT("DevToolAddress not set in, defaulting to 127.0.0.1:9999"));
		this->DevToolAddress = TEXT("127.0.0.1:9999");
	}

	// Read project name and version from config files
	FString productId;
	if (!GConfig->GetString(
		TEXT("OnlineSubsystemEpic"),
		TEXT("ProductId"),
		productId,
		GEngineIni))
	{
		UE_LOG_ONLINE(Warning, TEXT("Product Id is empty, add your product id from Epic Games DevPortal to the config files."));
		hasInvalidParams = true;
	}

	FString sandboxId;
	if (!GConfig->GetString(
		TEXT("OnlineSubsystemEpic"),
		TEXT("SandboxId"),
		sandboxId,
		GEngineIni))
	{
		UE_LOG_ONLINE(Warning, TEXT("Sandbox Id is empty, add your sandbox id from Epic Games DevPortal to the config files."));
		hasInvalidParams = true;
	}

	FString deploymentId;
	if (!GConfig->GetString(
		TEXT("OnlineSubsystemEpic"),
		TEXT("DeploymentId"),
		deploymentId,
		GEngineIni))
	{
		UE_LOG_ONLINE(Warning, TEXT("Deployment Id is empty, add your deployment id from Epic Games DevPortal to the config files."));
		hasInvalidParams = true;
	}

	FString clientId;
	GConfig->GetString(
		TEXT("OnlineSubsystemEpic"),
		TEXT("ClientCredentialsId"),
		clientId,
		GEngineIni
	);
	FString clientSecret;
	GConfig->GetString(
		TEXT("OnlineSubsystemEpic"),
		TEXT("ClientCredentialsSecret"),
		clientSecret,
		GEngineIni
	);
	if (clientId.IsEmpty() || clientSecret.IsEmpty())
	{
		UE_LOG_ONLINE(Warning, TEXT("[EOS SDK] Client credentials are invalid, check clientid and clientsecret!"));
		hasInvalidParams = true;
	}

	FString countryCode;
	char* countryCodeC = nullptr;
	if (GConfig->GetString(
		TEXT("OnlineSubsystemEpic"),
		TEXT("CountryCode"),
		countryCode,
		GEngineIni))
	{
		countryCodeC = TCHAR_TO_UTF8(*countryCode);
	}
	else
	{
		countryCodeC = nullptr;
	}

	FString localeCode;
	char* localeCodeC = nullptr;
	if (GConfig->GetString(
		TEXT("OnlineSubsystemEpic"),
		TEXT("LocaleCode"),
		localeCode,
		GEngineIni))
	{
		localeCodeC = TCHAR_TO_UTF8(*countryCode);
	}
	else
	{
		localeCodeC = nullptr;
	}

	FString encryptionKey;
	char* encryptionKeyC = nullptr;
	if (GConfig->GetString(
		TEXT("OnlineSubsystemEpic"),
		TEXT("EncryptionKey"),
		encryptionKey,
		GEngineIni))
	{
		if (encryptionKey.Len() != 64)
		{
			UE_LOG_ONLINE(Warning, TEXT("Got encryption key, but its length wasn't 64 characters."));
			hasInvalidParams = true;
		}
		else
		{
			encryptionKeyC = TCHAR_TO_UTF8(*encryptionKey);
		}
	}
	else
	{
		encryptionKeyC = nullptr;
	}
	
	FString cacheDirectory;
	char const* cacheDirectoryC = nullptr;
	if (GConfig->GetString(TEXT("OnlineSubsystemEpic"), TEXT("CacheDirectory"), cacheDirectory, GEngineIni)
		|| cacheDirectory.Len() == 0)
	{
		UE_LOG_ONLINE(Warning, TEXT("Got no cache directory, defaulting to %s"), *cacheDirectory);
		cacheDirectoryC = FUtils::GetTempDirectory();
	}
	else
	{
		cacheDirectoryC = TCHAR_TO_UTF8(*cacheDirectory);
	}

	uint64 platformFlags = 0;
#if UE_EDITOR
	// The platform overlay causes rendering artifacts in the editor,
	// this flag ensures it is not loaded in the editor or PIE
	platformFlags |= EOS_PF_LOADING_IN_EDITOR;
#else
	platformFlags |= 0;
#endif

	bool disableOverlay = false;
	if (GConfig->GetBool(TEXT("OnlineSubsystemEpic"), TEXT("DisableOverlay"), disableOverlay, GEngineIni))
	{
		if (disableOverlay)
		{
			platformFlags |= EOS_PF_DISABLE_OVERLAY;
		}
	}

	bool disableSocialOverlay = false;
	if (GConfig->GetBool(TEXT("OnlineSubsystemEpic"), TEXT("DisableSocialOverlay"), disableOverlay, GEngineIni))
	{
		if (disableSocialOverlay)
		{
			platformFlags |= EOS_PF_DISABLE_SOCIAL_OVERLAY;
		}
	}

	// The tick budget is set in the config as a double, not an unsigned 32-bit integer
	// This was done because UE has no way of retrieving an unsigned type from the config
	// and a 64 bit double has the same codomain as an unsigned 32-bit integer.
	double tickBudgetConfig = 0;
	uint32_t tickBudget = 0;
	if (!GConfig->GetDouble(TEXT("OnlineSubsystemEpic"), TEXT("TickBudget"), tickBudgetConfig, GEngineIni))
	{
		UE_LOG_ONLINE(Verbose, TEXT("No tick budget set, defaulting to 0"));
	}
	else
	{
		// Floor the value to the next value. This is a design choice and there is no right or wrong.
		// The reasoning is, that the user wants to use the tick budget at most and "stealing" more time
		// might result in less time for other tasks in the program.
		double d = FMath::FloorToDouble(tickBudgetConfig);

		// Convert to uint32_t by casting
		tickBudget = static_cast<uint32_t>(d);
	}

	if (hasInvalidParams)
	{
		return false;
	}


	// Create platform instance
	EOS_Platform_ClientCredentials clientCredentials = {
		TCHAR_TO_UTF8(*clientId),
		TCHAR_TO_UTF8(*clientSecret)
	};
	EOS_Platform_Options PlatformOptions = {
		EOS_PLATFORM_OPTIONS_API_LATEST,
		nullptr,					// MUST be nulled
		TCHAR_TO_UTF8(*productId),	// Required
		TCHAR_TO_UTF8(*sandboxId),	// Required
		clientCredentials,			// Required
		this->IsServer(),
		encryptionKeyC,
		countryCodeC,
		localeCodeC,
		TCHAR_TO_UTF8(*deploymentId), // Required
		platformFlags,
		cacheDirectoryC,
		tickBudget
	};
	this->PlatformHandle = EOS_Platform_Create(&PlatformOptions);
	if (!this->PlatformHandle)
	{
		UE_LOG_ONLINE(Warning, TEXT("[EOS SDK] Platform Create Failed!"));
		return false;
	}

	this->IdentityInterface = MakeShareable(new FOnlineIdentityInterfaceEpic(this));
	this->SessionInterface = MakeShareable(new FOnlineSessionEpic(this));
	this->UserInterface = MakeShareable(new FOnlineUserEpic(this));
	this->PresenceInterface = MakeShared<FOnlinePresenceEpic, ESPMode::ThreadSafe>(this);
	this->FriendsInterface = MakeShareable(new FOnlineFriendInterfaceEpic(this));

	this->IsInit = true;
	return true;
}

bool FOnlineSubsystemEpic::Shutdown()
{
	this->IsInit = false;
	this->PlatformHandle = nullptr;


#define DESTRUCT_INTERFACE(Interface) \
	if (Interface.IsValid()) \
	{ \
		ensure(Interface.IsUnique()); \
		Interface = nullptr; \
	}

	// Destruct the interfaces
	DESTRUCT_INTERFACE(IdentityInterface);
	DESTRUCT_INTERFACE(SessionInterface);
	DESTRUCT_INTERFACE(UserInterface);
	DESTRUCT_INTERFACE(this->PresenceInterface);
	DESTRUCT_INTERFACE(FriendsInterface);
	
#undef DESTRUCT_INTERFACE

	return true;
}

FString FOnlineSubsystemEpic::GetAppId() const
{
	// The AppId is a combination of the Projects id and a version in the form of:
	// {ProductId}::{ProductVersion}
	// This guarantees a unique identifier for a given product and version.
	// They however differ from the SDKs product id!

	FString projectId;
	GConfig->GetString(
		TEXT("/Script/EngineSettings.GeneralProjectSettings"),
		TEXT("ProjectId"),
		projectId,
		GGameIni
	);

	FString projectVersion;
	GConfig->GetString(
		TEXT("/Script/EngineSettings.GeneralProjectSettings"),
		TEXT("ProjectVersion"),
		projectVersion,
		GGameIni
	);

	return FString::Printf(TEXT("%s::%s"), *projectId, *projectVersion);
}

FText FOnlineSubsystemEpic::GetOnlineServiceName() const
{
	return NSLOCTEXT("OnlineSubsystemEpic", "OnlineServiceName", "Epic");
}

bool FOnlineSubsystemEpic::Tick(float DeltaTime)
{
	FOnlineSubsystemImpl::Tick(DeltaTime);

	if (this->PlatformHandle)
	{
		EOS_Platform_Tick(this->PlatformHandle);
	}

	if (this->SessionInterface)
	{
		this->SessionInterface->Tick(DeltaTime);
	}

	if (this->UserInterface)
	{
		this->UserInterface->Tick(DeltaTime);
	}

	return true;
}